File: src/main/scala/com/michalplachta/galactic/console/GalacticTwitterApp.scala
package com.michalplachta.galactic.console

import com.michalplachta.galactic.values.RemoteData._
import com.michalplachta.galactic.service.{ FollowersService, TweetsService }
import com.michalplachta.galactic.values.{ RemoteData, Tweet }

import scala.annotation.tailrec
import scala.io.StdIn

object GalacticTwitterApp extends App {
  @tailrec
  def runConsoleTwitter(): Unit = {
    println("---\nEnter Citizen's name: ")
    val name = StdIn.readLine()
    println(s"Getting followers for $name")
    val followersText = getFollowersText(name)
    val tweets = getTweetWall(name)
    println(s"$name has $followersText followers!")
    println(s"$name's Tweet Wall: $tweets")
    runConsoleTwitter()
  }

  def getFollowersText(citizenName: String): String = {
    val remoteFollowers: RemoteData[Int] = FollowersService.getFollowers(citizenName)
    remoteFollowers match {
      case NotRequestedYet()    ⇒ "(not requested yet)"
      case Loading()            ⇒ "(loading...)"
      case Fetched(count)       ⇒ count.toString
      case Failed(errorMessage) ⇒ s"(failed to get followers: $errorMessage)"
    }
  }

  def getTweetWall(citizenName: String): String = {
    val remoteTweets: RemoteData[List[Tweet]] = TweetsService.getTweetsFor(citizenName)
    remoteTweets match {
      case NotRequestedYet()    ⇒ "(not requested yet)"
      case Loading()            ⇒ "(loading...)"
      case Fetched(tweets)      ⇒ tweets.foldLeft("")((wall, t) ⇒ s"$wall\n${t.author.name}: ${t.text}")
      case Failed(errorMessage) ⇒ s"(failed to get tweets: $errorMessage)"
    }
  }

  runConsoleTwitter()
}

File: src/main/java/com/michalplachta/galactic/java/console/GalacticTwitterApp.java
package com.michalplachta.galactic.java.console;

import com.michalplachta.galactic.java.service.FollowersService;
import com.michalplachta.galactic.java.service.TweetsService;
import com.michalplachta.galactic.java.values.remotedata.RemoteData;
import com.michalplachta.galactic.java.values.Tweet;
import io.vavr.collection.List;

import java.util.Scanner;

import static com.michalplachta.galactic.java.internal.RemoteDataPatterns.*;
import static io.vavr.API.*;

public class GalacticTwitterApp {
    private static void runConsoleTwitter() {
        System.out.println("---\nEnter Citizen's name: ");
        Scanner scanner = new Scanner(System.in);
        String name = scanner.nextLine();
        System.out.println(String.format("Getting followers for %s", name));
        String followersText = getFollowersText(name);
        String tweets = getTweetWall(name);
        System.out.println(String.format("%s has %s followers!", name, followersText));
        System.out.println(String.format("%s's Tweet Wall: %s", name, tweets));
        runConsoleTwitter();
    }

    private static String getFollowersText(String citizenName) {
        RemoteData remoteFollowers = FollowersService.getFollowers(citizenName);
        return Match(remoteFollowers).of(
                Case(NotRequestedYet(), "(not requested yet)"),
                Case(Loading(), "(loading...)"),
                Case(Fetched($()), Object::toString),
                Case(Failed($()), errorMessage -> String.format("(failed to get followers: %s)", errorMessage))
        );
    }

    private static String getTweetWall(String citizenName) {
        RemoteData remoteTweets = TweetsService.getTweetsFor(citizenName);
        return Match(remoteTweets).of(
                Case(NotRequestedYet(), "(not requested yet)"),
                Case(Loading(), "(loading...)"),
                Case(Fetched($()), (List<Tweet> tweets) -> tweets.foldLeft("", (wall, t) -> String.format("%s\n%s: %s", wall, t.author.name, t.text))),
                Case(Failed($()), errorMessage -> String.format("(failed to get tweets: %s)", errorMessage))
        );
    }

    public static void main(String[] args) {
        runConsoleTwitter();
    }
}
